stellargraph/connector/neo4j/sampler.py

```diff
@@ -0,0 +1,202 @@
+                parameters={"node_id_list": cur_nodes, "num_samples": num_sample},
+            )
+            samples.append(result.data()[0]["next_samples"])
+
+        return samples
+
+
+@experimental(reason="the class is not fully tested")
+class Neo4JDirectedBreadthFirstNeighbors(GraphWalk):
+    """
+    Breadth First Walk that generates a sampled number of paths from a starting node.
+    It can be used to extract a random sub-graph starting from a set of initial nodes from Neo4J database.
+    """
+
+    def run(
+        self, neo4j_graphdb, nodes=None, n=1, in_size=None, out_size=None, seed=None
+    ):
+        """
+        Send queries to Neo4J databases and collect sampled breadth-first walks starting from the root nodes.
+
+        Args:
+            neo4j_graphdb (py2neo.Graph): the Neo4J Graph Database object
+            nodes (list): A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n (int): Number of walks per node id.
+            in_size (list): The number of in-directed nodes to sample with replacement at each depth of the walk.
+            out_size (list): The number of out-directed nodes to sample with replacement at each depth of the walk.
+            seed (int): Random number generator seed; default is None
+        Returns:
+            A list of multi-hop neighbourhood samples. Each sample expresses multiple undirected walks, but the in-node
+            neighbours and out-node neighbours are sampled separately. Each sample has the format:
+        """
+
+        def bfs_neighbor_query(sampling_direction):
+            direction = {"BOTH": "--", "IN": "<--", "OUT": "-->"}
+
+            return f"""
+                // expand the list of node id in seperate rows of ids.
+                UNWIND {{node_id_list}} AS node_id
+
+                // for each node id in every row, collect the random list of its neighbors.
+                CALL apoc.cypher.run(
+
+                    'MATCH(cur_node) WHERE id(cur_node) = {{node_id}}
+
+                    // find the neighbors
+                    MATCH (cur_node){direction[sampling_direction]}(neighbors)
+
+                    // put all ids into a list
+                    WITH CASE collect(id(neighbors)) WHEN [] THEN [null] ELSE collect(id(neighbors)) END AS in_neighbors_list
+
+                    // pick random nodes with replacement
+                    WITH apoc.coll.randomItems(in_neighbors_list, {{num_samples}}, True) AS in_samples_list
+
+                    RETURN in_samples_list',
+                    {{ node_id: node_id, num_samples: {{num_samples}}  }}) YIELD value
+
+                RETURN apoc.coll.flatten(collect(value.in_samples_list)) as next_samples
+            """
+
+        self._check_neighbourhood_sizes(in_size, out_size)
+        self._check_common_parameters(nodes, n, len(in_size), seed)
+
+        samples = [[head_node for head_node in nodes for _ in range(n)]]
```

Comment:
Repeating a list multiple times in Python can be done with `*`:

```suggestion
        samples = [nodes * n]
```

Reply:
`samples = [nodes * n]` will collate the lists n times. E.g `[1,2,3]*2` will be `[1,2,3,1,2,3]`. Would there be any nicer way to generate repeated items that are next to each other, i.e. `[1,1,2,2,3,3]`?

Reply:
Why do we need them to be next to each other?

Reply:
Thanks for your question. As I observed in the `SampledBreadthFirstWalk` and `DirectedBreadthFirstNeighbours`, walks from the same head node are next to each other when they are sampled repeatedly. 
At this moment, both ways do not hurt, as by default, we only sample once from each `head_node`.

Reply:
That's how the existing code is, yes, but I'm asking a slightly deeper question: do you think there's a reason that they're next to each other, or is it just coincidence?

Reply:
I personally think the main reason behind this arrangement amongst the head_nodes is to keep it consistent among other sampling algorithms (`UniformRandomMetaPathWalk`
, `TemporalRandomWalk`, etc.). 
I feel that it should be better if I can reproduce the result as similarly as we could for the original codes. But if there is any critical reason behind this, please let me know what you think.

Reply:
That sounds good to me; using `nodes * n` is only a tiny improvement so I agree with your assessment that it's not worth being different. If the code required to match the existing algorithms was more complicated, the decision might be different.